sound/soc/sof/intel/telemetry.c

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+
+#include <sound/sof/ipc4/header.h>
+#include <sound/sof/xtensa.h>
+#include "../ipc4-priv.h"
+#include "../sof-priv.h"
+#include "hda.h"
+#include "telemetry.h"
+
+static u32 find_telemetry_slots(struct snd_sof_dev *sdev)
+{
+	u32 slot_desc_type_offset;
+	u32 type;
+	int i;
+
+	slot_desc_type_offset = sdev->debug_box.offset + sizeof(u32);
+	for (i = 0; i < SOF_IPC4_MAX_DEBUG_SLOTS; i++) {
+		sof_mailbox_read(sdev, slot_desc_type_offset, &type, sizeof(type));
+
+		if (type == SOF_IPC4_DEBUG_SLOT_TELEMETRY)
+			return sdev->debug_box.offset + (i + 1) * SOF_IPC4_DEBUG_SLOT_SIZE;
+
+		/* The type is the second u32 in the slot descriptor */
+		slot_desc_type_offset += SOF_IPC4_DEBUG_DESCRIPTOR_SIZE;
+	}
+
+	dev_warn(sdev->dev, "Can't find telemetry in debug window\n");
+	return 0;
+}
+
+void sof_ipc4_intel_dump_telemetry_state(struct snd_sof_dev *sdev, u32 flags)
+{
+	char *level = (flags & SOF_DBG_DUMP_OPTIONAL) ? KERN_DEBUG : KERN_ERR;
+	struct sof_ipc_dsp_oops_xtensa xoops = { 0 };
+	void *telemetry_data;
+	u32 slot_offset;
+	u32 core;
+
+	slot_offset = find_telemetry_slots(sdev);
+	if (!slot_offset)
+		return;
+
+	telemetry_data = kmalloc(SOF_IPC4_DEBUG_SLOT_SIZE, GFP_KERNEL);
+	if (!telemetry_data)
+		return;
+
+	sof_mailbox_read(sdev, slot_offset, telemetry_data, SOF_IPC4_DEBUG_SLOT_SIZE);
+
+	core = FIELD_GET(SOF_DBG_DUMP_CORE_MASK, flags);
+	dev_err(sdev->dev, "Dsp panic on core %d\n", core);
+
+	sof_stack(sdev, level, &xoops, telemetry_data, SOF_IPC4_DEBUG_SLOT_SIZE);
```

Comment:
@RanderWang Wouldn't this parse the oops incorrectly as "telemetry_data" as the full telemetry window contents and not just the exception record?

Reply:
@kv2019i If we decide to parse exception record we need to get exact definition of telemetry_data so that we can get exact offset of exception record. In this way we have to deal with cSOF and REF FW in different way in kernel. Now I dump the full content of telemetry_data and let tool to support it. We can use different tools for different FW. This is the method used by windows driver.

Reply:
I think @RanderWang  that's ok, but doesn't sof_stack() except a specific layout of data, so the kernel messages that get printed are incorrect?

So if we want to export the raw telemetry dump, we'd need to use some other method to export it to user-space.

Some downsides in the generic approach:
  - we won't have user-readable summary of the DSP panic in kernel message
  - users cannot continue to use "exception" debugfs that is used by existing IPC3 users, see e.g. https://github.com/thesofproject/linux/issues/4274